wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/validator/DevfileIntegrityValidator.java

```diff
@@ -78,21 +79,33 @@ public void validateDevfile(Devfile devfile) throws DevfileFormatException {
           }
           editorTool = tool;
           break;
+
         case PLUGIN_TOOL_TYPE:
           // do nothing
           break;
+
         case KUBERNETES_TOOL_TYPE:
-        case OPENSHIFT_TOOL_TYPE:
           // fall through
+        case OPENSHIFT_TOOL_TYPE:
+          if (dockerimageTool != null) {
+            throw new DevfileFormatException(
+                "Devfile cannot contain kubernetes/openshift and dockerimage tool at the same time");
+          }
+          k8sOSTool = tool;
+          break;
+
         case DOCKERIMAGE_TOOL_TYPE:
-          if (recipeTool != null) {
+          if (k8sOSTool != null) {
```

Comment:
What is the root reason for not mixing kubernetes and docker tools?

Reply:
Because we store k8s/os/dockerimage tool in recipe and workspace can have only one recipe.
It is possible to generate Deployment for dockerimage tool and merge it with other k8s/os tools but it will be implemented in a separate issue.

Reply:
Is there an issue for it?

Reply:
Yes, https://github.com/eclipse/che/issues/12875